### PR TITLE
Woo: Refactor product variations edit reducer

### PR DESCRIPTION
**Pull request by coderkevin**
This refactors the first implementation of the product variations edit
reducer, which only supported the creation of variations. This now
supports the updating and deletion of variations. In addition, it will
automatically create and delete variations based on the variation types
available.

This PR also adds a lot of additional tests for the most likely scenarios of the product edit states.

To Test:
1. `npm test`
2. Visit store, Add a new product.
3. Make it a variable product and add variation types.
4. Observe the variation types and fill out some of the data there.

If that all works, there should be no regressions to the existing code.
Note that this PR doesn't contain the code required to add updates and deletes of product variations to the product action-list.

### CODE REVIEW

**Comment by coderkevin** [comment]
Rebased.

File: client/extensions/woocommerce/state/data-layer/ui/products/index.js [hunk 1]
```diff
@@ -23,15 +24,26 @@ import {
...
 export default {
+	[ WOOCOMMERCE_PRODUCT_EDIT ]: [ actionAppendProductVariations ],
+	[ WOOCOMMERCE_PRODUCT_ATTRIBUTE_EDIT ]: [ actionAppendProductVariations ],
 	[ WOOCOMMERCE_PRODUCT_CATEGORY_EDIT ]: [ handleProductCategoryEdit ],
 	[ WOOCOMMERCE_PRODUCT_ACTION_LIST_CREATE ]: [ handleProductActionListCreate ],
 };
 
+export function actionAppendProductVariations( { getState }, action ) {
+	const { siteId } = action;
+	const { id: productId } = action.product;
```
Comment (justinshreve) [inline client/extensions/woocommerce/state/data-layer/ui/products/index.js]: This line causes the first error, as `action.product` is `null` when I first load the form page.

File: client/extensions/woocommerce/state/ui/products/variations/test/edits-reducer.js [hunk 2]
```diff
@@ -2,23 +2,86 @@
...
+		name: 'New Variable Product',
+		attributes: [
+			{ uid: 'edit_0', name: 'Color', options: [ 'Black' ], variation: true },
+		],
+	};
+
+	const existingVariableProduct1 = {
+		id: 101,
+		type: 'variable',
+		name: 'Existing Simple Product',
```
Comment (coderkevin) [inline client/extensions/woocommerce/state/ui/products/variations/test/edits-reducer.js]: Thanks, got it.

File: client/extensions/woocommerce/state/ui/products/variations/test/edits-reducer.js [hunk 3]
```diff
@@ -2,23 +2,86 @@
...
+
+	const variationBlack = {
+		id: { index: 4 },
+		attributes: [
+			{ name: 'Color', option: 'Black' },
+		],
+	};
+
+	const variationBlue = {
+		id: { index: 4 },
```
Comment (coderkevin) [inline client/extensions/woocommerce/state/ui/products/variations/test/edits-reducer.js]: Reassigned.

**Comment by coderkevin** [comment]
Rebased and pulled some changes from later in the PR stack down to this PR to fix the issues mentioned. Please check commit 97c042a for details.

**Pull request merged**